### ISSUE TITLE
EP Plugin crashing with - Again: Resource temporarily unavailable

### ISSUE THREAD

**Issue by tomdee** [Metaswitch/calico-docker-prototype#22]
I'm seeing the endpoint plugin crash with the following error.

```
2015-01-21 02:39:14,652 [ERROR] __main__ 43: Starting up Docker demo endpoint plugin
2015-01-21 02:39:14,656 [DEBUG] __main__ 81: Read config from /config/data (['/config/data/192_168_1_3.txt', '/config/data/192_168_1_2.txt', '/config/data/192_168_1_1.txt', '/config/data/192_168_1_11.txt', '/config/data/192_168_1_10.txt', '/config/data/192_168_1_4.txt', '/config/data/felix.txt'])
2015-01-21 02:39:14,663 [DEBUG] __main__ 114:   Found configured endpoint 856a57733a1532a2ba211c4e8dabe4182c475e63c49005a9dbfc8d395de80684 (host=172_17_8_102, mac=d6:13:41:6b:c8:ea, ip=192.168.1.3, group=C_GROUP)
2015-01-21 02:39:14,663 [DEBUG] __main__ 114:   Found configured endpoint 5cde8f9fb40e0b8ace77410f762ebe777fea787774ac384157c4fe261315f46c (host=172_17_8_101, mac=7e:35:03:21:b2:d2, ip=192.168.1.2, group=SHARED_GROUP)
2015-01-21 02:39:14,668 [DEBUG] __main__ 114:   Found configured endpoint b0d882cdcaea2e0b10d61e1ffd920b3d79759a145b13cf1c1908aa6065694183 (host=172_17_8_101, mac=3a:2e:87:6a:9e:3c, ip=192.168.1.1, group=A_GROUP)
2015-01-21 02:39:14,670 [DEBUG] __main__ 114:   Found configured endpoint ea55dae198b3f5192c4eb91502014dcfd27e13eeefdb952400fe70a5127077ac (host=172_17_8_101, mac=a2:04:1b:20:36:e1, ip=192.168.1.11, group=GROUP)
2015-01-21 02:39:14,670 [DEBUG] __main__ 114:   Found configured endpoint cd445c9bbbf86bb1912c8db5d3caa029cb7a9258a27e3467134027d0e93fe2b3 (host=172_17_8_101, mac=b2:64:a0:47:65:24, ip=192.168.1.10, group=GROUP)
2015-01-21 02:39:14,671 [DEBUG] __main__ 114:   Found configured endpoint 50108efc51ccbc66300af25b1617b9fc15f9682de30a0de43edb9fbcd0c144de (host=172_17_8_102, mac=3a:45:97:0d:ef:94, ip=192.168.1.4, group=SHARED_GROUP)
2015-01-21 02:39:14,673 [DEBUG] __main__ 119:   Found configured Felix 172_17_8_101 at 172.17.8.101
2015-01-21 02:39:14,674 [DEBUG] __main__ 119:   Found configured Felix 172_17_8_102 at 172.17.8.102
2015-01-21 02:39:14,675 [DEBUG] __main__ 129: Created EP socket for resync
2015-01-21 02:39:19,684 [DEBUG] __main__ 81: Read config from /config/data (['/config/data/192_168_1_3.txt', '/config/data/192_168_1_2.txt', '/config/data/192_168_1_1.txt', '/config/data/192_168_1_11.txt', '/config/data/192_168_1_10.txt', '/config/data/192_168_1_4.txt', '/config/data/felix.txt'])
2015-01-21 02:39:19,686 [DEBUG] __main__ 114:   Found configured endpoint 856a57733a1532a2ba211c4e8dabe4182c475e63c49005a9dbfc8d395de80684 (host=172_17_8_102, mac=d6:13:41:6b:c8:ea, ip=192.168.1.3, group=C_GROUP)
2015-01-21 02:39:19,686 [DEBUG] __main__ 114:   Found configured endpoint 5cde8f9fb40e0b8ace77410f762ebe777fea787774ac384157c4fe261315f46c (host=172_17_8_101, mac=7e:35:03:21:b2:d2, ip=192.168.1.2, group=SHARED_GROUP)
2015-01-21 02:39:19,688 [DEBUG] __main__ 114:   Found configured endpoint b0d882cdcaea2e0b10d61e1ffd920b3d79759a145b13cf1c1908aa6065694183 (host=172_17_8_101, mac=3a:2e:87:6a:9e:3c, ip=192.168.1.1, group=A_GROUP)
2015-01-21 02:39:19,689 [DEBUG] __main__ 114:   Found configured endpoint ea55dae198b3f5192c4eb91502014dcfd27e13eeefdb952400fe70a5127077ac (host=172_17_8_101, mac=a2:04:1b:20:36:e1, ip=192.168.1.11, group=GROUP)
2015-01-21 02:39:19,690 [DEBUG] __main__ 114:   Found configured endpoint cd445c9bbbf86bb1912c8db5d3caa029cb7a9258a27e3467134027d0e93fe2b3 (host=172_17_8_101, mac=b2:64:a0:47:65:24, ip=192.168.1.10, group=GROUP)
2015-01-21 02:39:19,691 [DEBUG] __main__ 114:   Found configured endpoint 50108efc51ccbc66300af25b1617b9fc15f9682de30a0de43edb9fbcd0c144de (host=172_17_8_102, mac=3a:45:97:0d:ef:94, ip=192.168.1.4, group=SHARED_GROUP)
2015-01-21 02:39:19,692 [DEBUG] __main__ 119:   Found configured Felix 172_17_8_101 at 172.17.8.101
2015-01-21 02:39:19,698 [DEBUG] __main__ 119:   Found configured Felix 172_17_8_102 at 172.17.8.102
2015-01-21 02:39:19,699 [DEBUG] __main__ 178: Nothing happened - send lots of ENDPOINTCREATED messages to make sure
2015-01-21 02:39:19,700 [DEBUG] __main__ 215: Sending ENDPOINTCREATED to 172_17_8_101 : {'issued': 1421807959700, 'state': 'enabled', 'mac': '7e:35:03:21:b2:d2', 'addrs': [{'addr': '192.168.1.2'}], 'endpoint_id': '5cde8f9fb40e0b8ace77410f762ebe777fea787774ac384157c4fe261315f46c', 'resync_id': None, 'type': 'ENDPOINTCREATED'}
2015-01-21 02:39:24,379 [DEBUG] __main__ 218: Got endpoint created response
2015-01-21 02:39:24,379 [DEBUG] __main__ 215: Sending ENDPOINTCREATED to 172_17_8_101 : {'issued': 1421807964379, 'state': 'enabled', 'mac': 'b2:64:a0:47:65:24', 'addrs': [{'addr': '192.168.1.10'}], 'endpoint_id': 'cd445c9bbbf86bb1912c8db5d3caa029cb7a9258a27e3467134027d0e93fe2b3', 'resync_id': None, 'type': 'ENDPOINTCREATED'}
2015-01-21 02:39:28,126 [DEBUG] __main__ 218: Got endpoint created response
2015-01-21 02:39:28,127 [DEBUG] __main__ 215: Sending ENDPOINTCREATED to 172_17_8_101 : {'issued': 1421807968127, 'state': 'enabled', 'mac': '3a:2e:87:6a:9e:3c', 'addrs': [{'addr': '192.168.1.1'}], 'endpoint_id': 'b0d882cdcaea2e0b10d61e1ffd920b3d79759a145b13cf1c1908aa6065694183', 'resync_id': None, 'type': 'ENDPOINTCREATED'}
2015-01-21 02:39:32,355 [DEBUG] __main__ 218: Got endpoint created response
2015-01-21 02:39:32,355 [DEBUG] __main__ 215: Sending ENDPOINTCREATED to 172_17_8_101 : {'issued': 1421807972355, 'state': 'enabled', 'mac': 'a2:04:1b:20:36:e1', 'addrs': [{'addr': '192.168.1.11'}], 'endpoint_id': 'ea55dae198b3f5192c4eb91502014dcfd27e13eeefdb952400fe70a5127077ac', 'resync_id': None, 'type': 'ENDPOINTCREATED'}
2015-01-21 02:39:37,361 [ERROR] __main__ 319: Terminating on exception
Traceback (most recent call last):
  File "plugin.py", line 317, in <module>
    main()
  File "plugin.py", line 311, in main
    do_ep_api()
  File "plugin.py", line 180, in do_ep_api
    send_all_eps(create_sockets, host, None)
  File "plugin.py", line 217, in send_all_eps
    create_socket.recv()
  File "socket.pyx", line 622, in zmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:5560)
  File "socket.pyx", line 656, in zmq.backend.cython.socket.Socket.recv (zmq/backend/cython/socket.c:5380)
  File "socket.pyx", line 139, in zmq.backend.cython.socket._recv_copy (zmq/backend/cython/socket.c:1758)
  File "checkrc.pxd", line 15, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:5960)
Again: Resource temporarily unavailable
```

**Comment by Lukasa** [comment]
This crash comes from us sending an ENDPOINTCREATED to Felix and Felix not responding within the timeout (5 seconds). I think Felix logging would be good here to know for sure.

However, it probably wouldn't kill the plugin to give a better error message here either.

**Comment by ed-harrison** [comment]
Could we just increase the timeout for the prototype plug-in on the basis that
-  if its a real error, we could wait 10s to determine this
-  a proportion of people are likely to be running the demo prototype on virtualbox style systems which could reasonably be slow due to other things going on the host machine (i.e. we've seen this in our own testing more than once for this reason).

**Comment by plwhite** [comment]
Probably worth timing out after 10 seconds; it does look like some VMs are just very slow indeed.

**Comment by plwhite** [comment]
Fixed in PR #25
